From c9c2a92a066c1cda750d0abe00713b5a8165aea1 Mon Sep 17 00:00:00 2001
From: Chris Olin <contact@chrisolin.com>
Date: Mon, 5 Oct 2026 23:01:08 -0400
Subject: [PATCH] SetupWizard: Fit Welcome on short windows

The default welcome layout puts a weight-2 spacer above a full-width
logo. That drawable is 512dp by 260dp, so on a short window it fills
the viewport and the title and Start scroll off with no cue.

Windows at least 480dp tall keep that layout under layout-h480dp.
Shorter windows use a capped logo and a 32sp title. Tablets stay on
the smallest-width layout. Televisions use the same large composition
through a resource alias. The sw320dp-television alias exists because
available height would otherwise outrank UI mode.

Test: Welcome on a 640x480, 180dpi window, height 427dp
Change-Id: Ieeebc8a8ee3281e5feacfd4b72d91bd000f33908
Signed-off-by: Chris Olin <contact@chrisolin.com>
---
 res/layout-h480dp/welcome_activity.xml        | 132 ++++++++++++++++++
 res/layout-television/welcome_activity.xml    |  93 ------------
 res/layout/welcome_activity.xml               | 122 +++++++---------
 .../welcome_activity_large.xml}               |   0
 res/values-sw320dp-television/layouts.xml     |  13 ++
 res/values-sw600dp/layouts.xml                |  10 ++
 res/values-television/layouts.xml             |  10 ++
 .../setupwizard/WelcomeActivity.java          |  22 ++-
 8 files changed, 234 insertions(+), 168 deletions(-)
 create mode 100644 res/layout-h480dp/welcome_activity.xml
 delete mode 100644 res/layout-television/welcome_activity.xml
 rename res/{layout-sw600dp/welcome_activity.xml => layout/welcome_activity_large.xml} (100%)
 create mode 100644 res/values-sw320dp-television/layouts.xml
 create mode 100644 res/values-sw600dp/layouts.xml
 create mode 100644 res/values-television/layouts.xml

diff --git a/res/layout-h480dp/welcome_activity.xml b/res/layout-h480dp/welcome_activity.xml
new file mode 100644
index 0000000..97ea0b2
--- /dev/null
+++ b/res/layout-h480dp/welcome_activity.xml
@@ -0,0 +1,132 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     SPDX-FileCopyrightText: 2016 The CyanogenMod Project
+     SPDX-FileCopyrightText: 2017-2024 The LineageOS Project
+     SPDX-License-Identifier: Apache-2.0
+-->
+<com.google.android.setupdesign.GlifLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/setup_wizard_layout"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:orientation="vertical">
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="0dp"
+            android:layout_weight="1"
+            android:orientation="vertical"
+            android:gravity="center_vertical">
+
+            <Space
+                android:layout_width="match_parent"
+                android:layout_height="0dp"
+                android:layout_weight="2" />
+
+            <ImageView
+                android:id="@+id/brand_logo"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:adjustViewBounds="true"
+                android:scaleType="centerCrop"
+                android:src="@drawable/logo" />
+
+            <Space
+                android:layout_width="match_parent"
+                android:layout_height="0dp"
+                android:layout_weight="0.5" />
+        </LinearLayout>
+
+        <LinearLayout
+            android:layout_width="fill_parent"
+            android:layout_height="wrap_content"
+            android:layout_gravity="bottom"
+            android:orientation="vertical">
+
+            <LinearLayout
+                style="@style/SudContentFrame"
+                android:layout_width="fill_parent"
+                android:layout_height="wrap_content"
+                android:layout_gravity="bottom"
+                android:orientation="vertical"
+                android:paddingStart="@dimen/welcome_content_container_padding_start"
+                android:paddingEnd="@dimen/welcome_content_container_padding_end"
+                android:paddingBottom="@dimen/welcome_content_container_padding_bottom">
+
+                <TextView
+                    style="@style/WelcomeTitle"
+                    android:id="@+id/welcome_title"
+                    android:layout_width="fill_parent"
+                    android:layout_height="wrap_content"
+                    android:hyphenationFrequency="none"
+                    android:paddingBottom="@dimen/welcome_content_container_padding_bottom" />
+
+                <LinearLayout
+                    android:layout_width="fill_parent"
+                    android:layout_height="wrap_content"
+                    android:animateLayoutChanges="true"
+                    android:focusableInTouchMode="true"
+                    android:importantForAccessibility="no"
+                    android:orientation="vertical">
+
+                    <Button
+                        android:id="@+id/launch_accessibility"
+                        style="@style/SudGlifButton.Secondary"
+                        android:layout_width="fill_parent"
+                        android:layout_height="fill_parent"
+                        android:layout_gravity="center_vertical"
+                        android:drawableStart="@drawable/ic_eye"
+                        android:gravity="start|center"
+                        android:drawablePadding="@dimen/welcome_content_container_padding_end"
+                        android:textSize="@dimen/welcome_accessibility_text_size"
+                        android:textColor="?android:textColorPrimary"
+                        android:text="@string/accessibility_settings" />
+                </LinearLayout>
+            </LinearLayout>
+
+        </LinearLayout>
+
+        <LinearLayout
+            style="@style/SudGlifButtonBar.Stackable"
+            android:layout_width="fill_parent"
+            android:layout_height="wrap_content"
+            android:layout_gravity="bottom">
+
+            <Button
+                android:id="@+id/emerg_dialer"
+                style="@style/SudGlifButton.Secondary"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:text="@string/emergency_call" />
+
+            <Space
+                android:layout_width="0.0dip"
+                android:layout_height="0.0dip"
+                android:layout_weight="1.0" />
+
+            <Button
+                android:id="@+id/skip"
+                style="@style/SudGlifButton.Secondary"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:text="@string/skip"
+                android:visibility="gone" />
+
+            <Space
+                android:layout_width="0.0dip"
+                android:layout_height="0.0dip"
+                android:layout_weight="1.0" />
+
+            <Button
+                android:id="@id/start"
+                style="@style/SudGlifButton.Primary"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:text="@string/start" />
+        </LinearLayout>
+    </LinearLayout>
+</com.google.android.setupdesign.GlifLayout>
diff --git a/res/layout-television/welcome_activity.xml b/res/layout-television/welcome_activity.xml
deleted file mode 100644
index 2ea0a5e..0000000
--- a/res/layout-television/welcome_activity.xml
+++ /dev/null
@@ -1,93 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!--
-     SPDX-FileCopyrightText: 2016 The CyanogenMod Project
-     SPDX-FileCopyrightText: 2017-2022 The LineageOS Project
-     SPDX-License-Identifier: Apache-2.0
--->
-<LinearLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
-    android:orientation="vertical"
-    android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    style="@style/SudContentFrame">
-
-    <LinearLayout
-        android:layout_width="match_parent"
-        android:layout_height="match_parent"
-        android:gravity="center_vertical"
-        android:weightSum="100"
-        android:orientation="vertical">
-
-        <ImageView
-            android:id="@+id/brand_logo"
-            android:layout_width="wrap_content"
-            android:layout_height="0dp"
-            android:layout_weight="65"
-            android:layout_gravity="center"
-            android:src="@drawable/logo"
-            android:adjustViewBounds="true"
-            android:scaleType="fitCenter"
-            android:gravity="center"
-            android:paddingTop="@dimen/content_margin_top"
-            android:paddingRight="@dimen/content_margin_left"
-            android:paddingLeft="@dimen/content_margin_left"
-            android:paddingBottom="0dp" />
-
-        <LinearLayout
-            android:layout_width="match_parent"
-            android:layout_height="0dp"
-            android:layout_weight="35"
-            android:gravity="center_vertical"
-            android:orientation="vertical">
-
-            <TextView
-                style="@style/WelcomeTitle"
-                android:id="@+id/welcome_title"
-                android:layout_width="match_parent"
-                android:layout_height="wrap_content"
-                android:hyphenationFrequency="none"
-                android:gravity="center" />
-
-            <Button
-                android:id="@+id/start"
-                android:layout_width="wrap_content"
-                android:layout_height="wrap_content"
-                android:layout_gravity="center"
-                android:text="@string/start"
-                style="@style/SudGlifButton.Primary" />
-        </LinearLayout>
-
-    </LinearLayout>
-
-    <LinearLayout
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:orientation="horizontal"
-        android:gravity="center"
-        android:visibility="gone">
-
-        <Button
-            android:id="@+id/emerg_dialer"
-            android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
-            android:drawableTop="@drawable/ic_dialer"
-            android:text="@string/emergency_call"
-            style="@style/SudGlifButton.Secondary" />
-
-        <Button
-            android:id="@+id/launch_accessibility"
-            android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
-            android:drawableTop="@drawable/ic_eye"
-            android:text="@string/accessibility_settings"
-            style="@style/SudGlifButton.Secondary" />
-
-        <Button
-            android:id="@+id/skip"
-            style="@style/SudGlifButton.Secondary"
-            android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
-            android:text="@string/skip"
-            android:visibility="gone" />
-    </LinearLayout>
-</LinearLayout>
diff --git a/res/layout/welcome_activity.xml b/res/layout/welcome_activity.xml
index 97ea0b2..e9ea3b0 100644
--- a/res/layout/welcome_activity.xml
+++ b/res/layout/welcome_activity.xml
@@ -1,9 +1,15 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
      SPDX-FileCopyrightText: 2016 The CyanogenMod Project
-     SPDX-FileCopyrightText: 2017-2024 The LineageOS Project
+     SPDX-FileCopyrightText: 2017-2026 The LineageOS Project
      SPDX-License-Identifier: Apache-2.0
 -->
+<!--
+     Short windows. The weight-2 spacer and the 512dp by 260dp logo fill
+     the viewport, so the title and Start scroll off with no cue. Windows
+     at least 480dp tall keep the previous layout. Tablets and televisions
+     use welcome_activity_large.
+-->
 <com.google.android.setupdesign.GlifLayout
     xmlns:android="http://schemas.android.com/apk/res/android"
     android:id="@+id/setup_wizard_layout"
@@ -13,88 +19,56 @@
     <LinearLayout
         android:layout_width="match_parent"
         android:layout_height="match_parent"
+        android:gravity="center_horizontal"
         android:orientation="vertical">
 
+        <ImageView
+            android:id="@+id/brand_logo"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="8dp"
+            android:adjustViewBounds="true"
+            android:maxWidth="220dp"
+            android:maxHeight="112dp"
+            android:scaleType="fitCenter"
+            android:src="@drawable/logo" />
+
         <LinearLayout
+            style="@style/SudContentFrame"
             android:layout_width="match_parent"
-            android:layout_height="0dp"
-            android:layout_weight="1"
+            android:layout_height="wrap_content"
             android:orientation="vertical"
-            android:gravity="center_vertical">
+            android:paddingStart="16dp"
+            android:paddingEnd="16dp"
+            android:paddingBottom="8dp">
 
-            <Space
-                android:layout_width="match_parent"
-                android:layout_height="0dp"
-                android:layout_weight="2" />
-
-            <ImageView
-                android:id="@+id/brand_logo"
+            <TextView
+                style="@style/WelcomeTitle"
+                android:id="@+id/welcome_title"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:adjustViewBounds="true"
-                android:scaleType="centerCrop"
-                android:src="@drawable/logo" />
+                android:hyphenationFrequency="none"
+                android:lineHeight="40dp"
+                android:paddingBottom="8dp"
+                android:textSize="32sp" />
 
-            <Space
+            <Button
+                android:id="@+id/launch_accessibility"
+                style="@style/SudGlifButton.Secondary"
                 android:layout_width="match_parent"
-                android:layout_height="0dp"
-                android:layout_weight="0.5" />
-        </LinearLayout>
-
-        <LinearLayout
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_gravity="bottom"
-            android:orientation="vertical">
-
-            <LinearLayout
-                style="@style/SudContentFrame"
-                android:layout_width="fill_parent"
                 android:layout_height="wrap_content"
-                android:layout_gravity="bottom"
-                android:orientation="vertical"
-                android:paddingStart="@dimen/welcome_content_container_padding_start"
-                android:paddingEnd="@dimen/welcome_content_container_padding_end"
-                android:paddingBottom="@dimen/welcome_content_container_padding_bottom">
-
-                <TextView
-                    style="@style/WelcomeTitle"
-                    android:id="@+id/welcome_title"
-                    android:layout_width="fill_parent"
-                    android:layout_height="wrap_content"
-                    android:hyphenationFrequency="none"
-                    android:paddingBottom="@dimen/welcome_content_container_padding_bottom" />
-
-                <LinearLayout
-                    android:layout_width="fill_parent"
-                    android:layout_height="wrap_content"
-                    android:animateLayoutChanges="true"
-                    android:focusableInTouchMode="true"
-                    android:importantForAccessibility="no"
-                    android:orientation="vertical">
-
-                    <Button
-                        android:id="@+id/launch_accessibility"
-                        style="@style/SudGlifButton.Secondary"
-                        android:layout_width="fill_parent"
-                        android:layout_height="fill_parent"
-                        android:layout_gravity="center_vertical"
-                        android:drawableStart="@drawable/ic_eye"
-                        android:gravity="start|center"
-                        android:drawablePadding="@dimen/welcome_content_container_padding_end"
-                        android:textSize="@dimen/welcome_accessibility_text_size"
-                        android:textColor="?android:textColorPrimary"
-                        android:text="@string/accessibility_settings" />
-                </LinearLayout>
-            </LinearLayout>
-
+                android:drawableStart="@drawable/ic_eye"
+                android:drawablePadding="16dp"
+                android:gravity="start|center_vertical"
+                android:text="@string/accessibility_settings"
+                android:textColor="?android:textColorPrimary"
+                android:textSize="16sp" />
         </LinearLayout>
 
         <LinearLayout
             style="@style/SudGlifButtonBar.Stackable"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_gravity="bottom">
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content">
 
             <Button
                 android:id="@+id/emerg_dialer"
@@ -104,9 +78,9 @@
                 android:text="@string/emergency_call" />
 
             <Space
-                android:layout_width="0.0dip"
-                android:layout_height="0.0dip"
-                android:layout_weight="1.0" />
+                android:layout_width="0dp"
+                android:layout_height="0dp"
+                android:layout_weight="1" />
 
             <Button
                 android:id="@+id/skip"
@@ -117,9 +91,9 @@
                 android:visibility="gone" />
 
             <Space
-                android:layout_width="0.0dip"
-                android:layout_height="0.0dip"
-                android:layout_weight="1.0" />
+                android:layout_width="0dp"
+                android:layout_height="0dp"
+                android:layout_weight="1" />
 
             <Button
                 android:id="@id/start"
diff --git a/res/layout-sw600dp/welcome_activity.xml b/res/layout/welcome_activity_large.xml
similarity index 100%
rename from res/layout-sw600dp/welcome_activity.xml
rename to res/layout/welcome_activity_large.xml
diff --git a/res/values-sw320dp-television/layouts.xml b/res/values-sw320dp-television/layouts.xml
new file mode 100644
index 0000000..d26d122
--- /dev/null
+++ b/res/values-sw320dp-television/layouts.xml
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     SPDX-FileCopyrightText: 2016 The CyanogenMod Project
+     SPDX-FileCopyrightText: 2017-2026 The LineageOS Project
+     SPDX-License-Identifier: Apache-2.0
+-->
+<!--
+     Smallest width outranks available height, and UI mode does not.
+     A television at least 480dp tall would otherwise take layout-h480dp.
+-->
+<resources>
+    <item name="welcome_activity" type="layout">@layout/welcome_activity_large</item>
+</resources>
diff --git a/res/values-sw600dp/layouts.xml b/res/values-sw600dp/layouts.xml
new file mode 100644
index 0000000..ee736d1
--- /dev/null
+++ b/res/values-sw600dp/layouts.xml
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     SPDX-FileCopyrightText: 2016 The CyanogenMod Project
+     SPDX-FileCopyrightText: 2017-2026 The LineageOS Project
+     SPDX-License-Identifier: Apache-2.0
+-->
+<!-- Tablets keep the large welcome composition by smallest width. -->
+<resources>
+    <item name="welcome_activity" type="layout">@layout/welcome_activity_large</item>
+</resources>
diff --git a/res/values-television/layouts.xml b/res/values-television/layouts.xml
new file mode 100644
index 0000000..af0754a
--- /dev/null
+++ b/res/values-television/layouts.xml
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     SPDX-FileCopyrightText: 2016 The CyanogenMod Project
+     SPDX-FileCopyrightText: 2017-2026 The LineageOS Project
+     SPDX-License-Identifier: Apache-2.0
+-->
+<!-- Televisions keep the large welcome composition. -->
+<resources>
+    <item name="welcome_activity" type="layout">@layout/welcome_activity_large</item>
+</resources>
diff --git a/src/org/lineageos/setupwizard/WelcomeActivity.java b/src/org/lineageos/setupwizard/WelcomeActivity.java
index 9d6765e..01a99b4 100644
--- a/src/org/lineageos/setupwizard/WelcomeActivity.java
+++ b/src/org/lineageos/setupwizard/WelcomeActivity.java
@@ -1,6 +1,6 @@
 /*
  * SPDX-FileCopyrightText: 2016 The CyanogenMod Project
- * SPDX-FileCopyrightText: 2017-2024 The LineageOS Project
+ * SPDX-FileCopyrightText: 2017-2026 The LineageOS Project
  * SPDX-License-Identifier: Apache-2.0
  */
 
@@ -9,6 +9,7 @@ package org.lineageos.setupwizard;
 import static org.lineageos.setupwizard.SetupWizardApp.ACTION_EMERGENCY_DIAL;
 
 import android.content.Intent;
+import android.content.res.Configuration;
 import android.os.Bundle;
 import android.os.Build;
 import android.view.View;
@@ -25,6 +26,12 @@ public class WelcomeActivity extends SubBaseActivity {
     private static final String ACTION_ACCESSIBILITY_SETTINGS =
             "android.settings.ACCESSIBILITY_SETTINGS_FOR_SUW";
 
+    // The default welcome composition needs a tall window. Below this height the
+    // Glif header is an empty band, because Welcome sets no header title. Leave
+    // it in place at the 600dp two-pane width, where that header is the left pane.
+    private static final int SHORT_WELCOME_HEIGHT_DP = 480;
+    private static final int TWO_PANE_WIDTH_DP = 600;
+
     @Override
     protected void onStartSubactivity() {
     }
@@ -32,6 +39,7 @@ public class WelcomeActivity extends SubBaseActivity {
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        hideEmptyHeaderOnShortWindow();
         onSetupStart();
         SystemBarHelper.setBackButtonVisible(getWindow(), false);
         setNextText(R.string.start);
@@ -75,6 +83,18 @@ public class WelcomeActivity extends SubBaseActivity {
     public void onBackPressed() {
     }
 
+    private void hideEmptyHeaderOnShortWindow() {
+        Configuration config = getResources().getConfiguration();
+        if (config.screenHeightDp >= SHORT_WELCOME_HEIGHT_DP
+                || config.screenWidthDp >= TWO_PANE_WIDTH_DP) {
+            return;
+        }
+        View header = findViewById(com.google.android.setupdesign.R.id.sud_layout_header);
+        if (header != null) {
+            header.setVisibility(View.GONE);
+        }
+    }
+
     @Override
     protected int getLayoutResId() {
         return R.layout.welcome_activity;
-- 
2.55.0

